api/envoy/config/listener/v3/listener.proto

```diff
@@ -120,6 +121,76 @@ message Listener {
+  //
+  // Example 1: The following matcher selects three filter chains as follows:
+  //
+  // * if the destination port is 80, then the filter chain "http" is selected;
+  // * if the destination port is 443 and the source IP is in the range 192.0.0.0/2, then the filter chain "internal" is selected;
+  // * otherwise, if the destination port is 443, then the filter chain "https" is selected;
+  // * otherwise, the default filter chain is selected (or the connection is rejected without the default filter chain).
+  //
+  // .. validated-code-block:: yaml
+  //   :type-name: envoy.config.listener.v3.Listener
+  //
+  //   filter_chain_matcher:
+  //     matcher_tree:
+  //       input:
+  //         name: port
+  //         typed_config:
+  //           "@type": type.googleapis.com/envoy.extensions.matching.common_inputs.network.v3.DestinationPortInput
+  //       exact_match_map:
+  //         map:
+  //           "80":
+  //             action:
+  //               name: http
+  //               typed_config:
+  //                 "@type": type.googleapis.com/google.protobuf.StringValue
+  //                 value: http
+  //           "443":
+  //             matcher:
+  //               matcher_tree:
+  //                 input:
+  //                   name: ip
+  //                   typed_config:
+  //                     "@type": type.googleapis.com/envoy.extensions.matching.common_inputs.network.v3.SourceIPInput
+  //                 custom_match:
+  //                   name: ip-matcher
+  //                   typed_config:
+  //                     "@type": type.googleapis.com/xds.type.matcher.v3.IPMatcher
+  //                     range_matchers:
+  //                     - ranges:
+  //                       - address_prefix: 192.0.0.0
+  //                         prefix_len: 2
+  //                       on_match:
+  //                         action:
+  //                           name: internal
+  //                           typed_config:
+  //                             "@type": type.googleapis.com/google.protobuf.StringValue
+  //                             value: internal
+  //                     - ranges:
+  //                       - address_prefix: 0.0.0.0
+  //                       on_match:
+  //                         action:
+  //                           name: https
+  //                           typed_config:
+  //                             "@type": type.googleapis.com/google.protobuf.StringValue
+  //                             value: https
+  //
+  // .. note::
+  //
+  //  Once matched, each connection is permanently bound to its filter chain.
+  //  If the matcher changes but the filter chain remains the same, the
+  //  connections bound to the filter chain are not drained. If, however, the
+  //  filter chain is removed or structurally modified, then the drain for its
+  //  connections is initiated.
+  xds.type.matcher.v3.Matcher filter_chain_matcher = 32;
```

Comment:
A few doc notes:
1) We will need a release note.
2) I feel strongly that we have to have a) documentation for each supported matcher type and b) an auto generated list of all of the registered matcher types that will work.

I apologize in advance for putting this on, but we are going to need to sort out some of the documentation issues with putting protos in the xDS repo. We will need to make sure we are running protodoc on that repo, and also will need to figure out a way to wire up that repo to our extension category documentation tooling. If you need more pointers let me know but I might advise contacting @phlax on Slack and scheduling a short meeting with him to come up with a plan here. Again, I'm sorry to put this on you, but we need to figure this out because we will accrue significant documentation debt by not doing this, and the more protos that go into the xDS repo the worse this will get. Thank you in advance.

Separately, thoughts on marking this WiP for some period of time just so we can get some bake on this before marking it production ready and locked for API and security?

Reply:
Ack on making this WIP. I'll tag the protos as such and add a release note about experimental status for now. For documentation of cncf/xds, I'll ping @phlax and probably @markdroth. For matcher documentation, we should be able to figure out a table for inputs (TCP vs HTTP), and a list of all matchers using the existing extension categories.